src/main.rs

```diff
@@ -89,51 +97,65 @@ fn parse_args() -> Action {
     }
 }
 
-fn main() {
+fn run_app() -> Result<ReturnCode, error::Error> {
     let action = parse_args();
     match action {
         Action::Build { db_path, dir_path, threads } => {
-            let database = Database::build(&dir_path, true, threads).unwrap();
+            let database = Database::build(&dir_path, true, threads)?;
 
             {
                 let mut json_path = PathBuf::from(&db_path);
                 json_path.set_extension("json");
-                database.dump_json(json_path).unwrap();
+                database.dump_json(json_path)?;
             }
 
             {
                 let mut cbor_path = PathBuf::from(&db_path);
                 cbor_path.set_extension("cbor");
-                database.dump_cbor(cbor_path).unwrap();
+                database.dump_cbor(cbor_path)?;
             }
 
             {
                 let mut msgpack_path = PathBuf::from(&db_path);
                 msgpack_path.set_extension("msgpack");
-                database.dump_msgpack(msgpack_path).unwrap();
+                database.dump_msgpack(msgpack_path)?;
             }
 
-            let json = serde_json::to_string(&database).unwrap();
+            let json = serde_json::to_string(&database)?;
             println!("JSON bytes: {}", json.len());
-            let cbor = serde_cbor::to_vec(&database).unwrap();
+            let cbor = serde_cbor::to_vec(&database)?;
             println!("CBOR bytes: {}", cbor.len());
-            let msgpack = rmp_serde::to_vec(&database).unwrap();
+            let msgpack = rmp_serde::to_vec(&database)?;
             println!("MsgPack bytes: {}", msgpack.len());
+            Ok(ReturnCode::Built)
         }
         Action::Check { db_path, dir_path, threads } => {
             let mut cbor_path = PathBuf::from(&db_path);
             cbor_path.set_extension("cbor");
-            let database = Database::load_cbor(&cbor_path).unwrap();
-            database.check(&dir_path, threads).unwrap();
+            let database = Database::load_cbor(&cbor_path)?;
+            Ok(ReturnCode::Succeeded(database.check(&dir_path, threads)?))
         }
         Action::Diff { old_path, new_path } => {
             let mut cbor_old_path = PathBuf::from(&old_path);
             cbor_old_path.set_extension("cbor");
             let mut cbor_new_path = PathBuf::from(&new_path);
             cbor_new_path.set_extension("cbor");
-            let old = Database::load_cbor(&cbor_old_path).unwrap();
-            let new = Database::load_cbor(&cbor_new_path).unwrap();
-            old.show_diff(&new);
+            let old = Database::load_cbor(&cbor_old_path)?;
+            let new = Database::load_cbor(&cbor_new_path)?;
+            Ok(ReturnCode::Succeeded(old.show_diff(&new)))
         }
     }
 }
+
+fn main() {
+    ::std::process::exit(match run_app() {
+       Ok(return_code) => match return_code {
+           ReturnCode::Succeeded(diff_code) => diff_code as i32,
+           _ => 0,
```

Comment:
This is the place where we encode the return code. I'd explicitly enumerate the cases:

```
Build => 0,
Diff(NoChanges) => 0,
Diff(Changes) => 1,
Diff(Suspicious) => 2,
```

(Except with real syntax.)